### PR TITLE
[SPARK-11453][SQL] append data to partitioned table will messes up the result

### PR DESCRIPTION
**Pull request by cloud-fan**
The reason is that: 
1. For partitioned hive table, we will move the partitioned columns after data columns. (e.g. `<a: Int, b: Int>` partition by `a` will become `<b: Int, a: Int>`)
2. When append data to table, we use position to figure out how to match input columns to table's columns.

So when we append data to partitioned table, we will match wrong columns between input and table. A solution is reordering the input columns before match by position, like what we did for [`InsertIntoHadoopFsRelation`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelation.scala#L101-L105)

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai

**Comment by cloud-fan** [comment]
cc @liancheng

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 1]
```diff
@@ -169,11 +169,21 @@ final class DataFrameWriter private[sql](df: DataFrame) {
   private def insertInto(tableIdent: TableIdentifier): Unit = {
     val partitions = partitioningColumns.map(_.map(col => col -> (None: Option[String])).toMap)
     val overwrite = mode == SaveMode.Overwrite
+
+    // A partitioned relation schema's can be different from the input logicalPlan, since
+    // partition columns are all moved after data column. We Project to adjust the ordering.
+    // TODO: this belongs in the analyzer.
+    val input = partitioningColumns.map { parCols =>
+      val projectList = df.logicalPlan.output.filterNot(c => parCols.contains(c.name)) ++
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Do we need to consider case sensitivity for partition column names here?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 2]
```diff
@@ -169,11 +169,21 @@ final class DataFrameWriter private[sql](df: DataFrame) {
   private def insertInto(tableIdent: TableIdentifier): Unit = {
     val partitions = partitioningColumns.map(_.map(col => col -> (None: Option[String])).toMap)
     val overwrite = mode == SaveMode.Overwrite
+
+    // A partitioned relation schema's can be different from the input logicalPlan, since
+    // partition columns are all moved after data column. We Project to adjust the ordering.
+    // TODO: this belongs in the analyzer.
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Nit: some typos

- "... relation schema's ..." => "... relation's schema ..."
- "... after data column." => "... after data columns"
- "... belongs in ..." => "... belongs to ..."

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 3]
```diff
@@ -169,11 +169,21 @@ final class DataFrameWriter private[sql](df: DataFrame) {
...
+    val input = partitioningColumns.map { parCols =>
+      val projectList = df.logicalPlan.output.filterNot(c => parCols.contains(c.name)) ++
+        parCols.map(UnresolvedAttribute(_))
+      Project(projectList, df.logicalPlan)
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I found this block is a little bit hard to understand. The following version may be questionably more readable:

```scala
      val (inputDataCols, inputPartCols) = df.logicalPlan.output.partition {
        c => partCols.contains(c.name)
      }
      val projectList = inputDataCols ++ inputPartCols.map(c => UnresolvedAttribute(c.name))
      Project(projectList, df.logicalPlan)
```

**Comment by liancheng** [comment]
Overall looks good, left some minor comments.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 4]
```diff
@@ -167,17 +167,39 @@ final class DataFrameWriter private[sql](df: DataFrame) {
...
         ifNotExists = false)).toRdd
   }
 
+  private def normalizedParCols: Option[Seq[String]] = partitioningColumns.map { parCols =>
+    parCols.map { col =>
+      df.logicalPlan.output
+        .map(_.name)
+        .find(df.queryExecution.analyzer.resolver(_, col))
+        .getOrElse(throw new AnalysisException(
+          s"Partition column $col not found in schema ${df.logicalPlan.schema}"))
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: We only need to print column names, right?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 5]
```diff
@@ -167,17 +167,39 @@ final class DataFrameWriter private[sql](df: DataFrame) {
...
     val overwrite = mode == SaveMode.Overwrite
+
+    // A partitioned relation's schema can be different from the input logicalPlan, since
+    // partition columns are all moved after data columns. We Project to adjust the ordering.
+    // TODO: this belongs to the analyzer.
+    val input = normalizedParCols.map { parCols =>
+      val (inputPartCols, inputDataCols) = df.logicalPlan.output.partition { attr =>
+        parCols.contains(attr.name)
+      }
+      val projectList = inputDataCols ++ inputPartCols.map(c => UnresolvedAttribute(c.name))
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Do we need to make them unresolved?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 6]
```diff
@@ -167,17 +167,38 @@ final class DataFrameWriter private[sql](df: DataFrame) {
...
     val overwrite = mode == SaveMode.Overwrite
+
+    // A partitioned relation's schema can be different from the input logicalPlan, since
+    // partition columns are all moved after data columns. We Project to adjust the ordering.
+    // TODO: this belongs to the analyzer.
+    val input = normalizedParCols.map { parCols =>
+      val (inputPartCols, inputDataCols) = df.logicalPlan.output.partition { attr =>
+        parCols.contains(attr.name)
+      }
+      Project(inputDataCols ++ inputPartCols, df.logicalPlan)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: actually, can we have a check to make sure that partition columns do appear at the end of the column list?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: hmmm,  I think `normalizedParCols` already did this check?

**Comment by cloud-fan** [comment]
retest this please

**Comment by yhuai** [comment]
LGTM. Merging to master and branch 1.6.

**Pull request closed without merging**